### ISSUE TITLE
深度学习模型组成

### ISSUE THREAD

**Issue by penguincwarrior** [penguincwarrior/penguincwarrior#4]
1. Model

1.1 CNN

1.1.1 CNN models
- basic（with he Initialization）
- LeNet
- AlexNet
- VGGNet
- DenseNet
- ResNet
思想：对于深层网络，如果其N层之后都在做恒等变换，那么其性能应该不差于N层的浅层网络。
![](https://pic1.zhimg.com/80/v2-bc88ce9d1f45720932dc4db9f5b03608_hd.jpg)
![image](https://user-images.githubusercontent.com/3872795/71085230-5484f780-2197-11ea-8fa4-29a98d790312.png)
1）通过在反向传播中加入常数1（在网络中加入输入x）保证残差网络的梯度传播不会爆炸或消失；
2）打破网络对称性，从而强迫网络对输入敏感
当层数为1时，加入shortcut不会提高性能。
当输出输出维度不同时，使用1*1卷积调整x的维度。
https://zhuanlan.zhihu.com/p/42833949
- GoogleNet
通过Inception将多个尺寸的卷积层和pooling层（都具有相同的stride）集成到一层中，然后进行聚合。

1.1.2 CNN layers
Convolutional layer：小的kernel size（2-3），逐渐增加的channel size（1.5～2x），多个conv直接连接（例如两个3*3卷积层的连接等价于5*5的卷积层），逐渐增加的直接连接的conv层
1*1卷积核仅改变channel数量，因此可以用来升维或降维，一般用来1）对齐不同sibling的channel数；2）通过降维降低后续计算（如5*5conv）开销。
设计原则：1）当输出维度相同时，filter数目相同；2）当输出维度降低X时，filter增加X
pooling: maxpooling, average pooling(用于取代最后的全连接，或者在全连接钱使用)

1.2 Network in Network

1.3 Siamese Network

1.4 AutoEncoder

- basic
- convolutional autoencoder
- variational autoencoder
- conditional variational autoencoder

1.5 GAN

- basic GAN
- Wasserstein GAN
- Convolutional GAN
- Convolutional Wasserstein GAN

1.6 GNN

- basic
- GGNN
- GAT
- GCN

1.7 RNN

- basic
- LSTM
- GRU
- bi-directional RNN
- seq2seq

1.8 ordinal regression

- ordinal regression CNN

2. Transfer Learning

3. Tricks

3.1 Initialization:

- random initialization(for shallow networks, 11 layers, for example, 0 mean, 10-2 variance)
- he Initialization
- reuse shallow layers(first 4 conv layers and fc layers)
- Glorot initialization

3.2 Layers:

- Batch Normalization（在conv和激活函数中间使用）
我的理解是，在早期layer-by-layer的训练方式中，当前训练层的输入分布是稳定的，因此能够稳定地学习到权重向量。而在目前通用的学习范式中，由于上一层的输出一直在更新，可能出现权重过大或过小等问题，从而难以保证下层训练的稳定性，即covariation shift。Batch Normalization是这个问题的解决方案。
对于每个隐层神经元，把逐渐向非线性函数映射后向取值区间极限饱和区靠拢的输入分布强制拉回到均值为0方差为1的比较标准的正态分布，使得非线性变换函数的输入值落入对输入比较敏感的区域，以此避免梯度消失问题（https://www.cnblogs.com/guoyaohua/p/8724433.html）
问题：
1）什么不直接把上层权重标准化为mean=0，var=1的标准分布？核心思想应该是想找到一个线性和非线性的较好平衡点，既能享受非线性的较强表达能力的好处，又避免太靠非线性区两头使得网络收敛速度太慢。（https://www.cnblogs.com/guoyaohua/p/8724433.html）
2）BN只对训练效率有影响吗？还是对精度也有影响？主要对训练效率有影响，但在有限的计算资源中可以进行更多的超参数调整次数，从而更可能增加模型精度；对模型精度的影响集中在防止模型过拟合上。
3）训练集和测试集的BN不同吗？在测试集中，BN中的均值和方差不使用测试样本的统计量，而使用训练集X总体的无偏统计量。
- layer normalization
和batch normalization相比：1）针对mini-batch进行优化；2）解决RNN下BN构建问题。
使用场景：mini-batch很小或使用RNN，否则尤其对于CNN还是使用BN更加友好
- dropout(0.5)
防止神经元间的co-adaptation

3.3 Optimization:

- Adam
- momentum(0.9)
- SGD
- mini-batch（256）
    - Annealing with Increasing the Batch Size
- Cyclical Learning Rate
- Gradient Clipping
- manually adjust learning rate (/10 when no improvement occurs；×0.96 for every 8 epochs)

3.4 hyperparameter tuning

3.5 activation function
ReLU
sigmoid
tanh

3.6 Regularization
- L2(5 · 10−4，10-4)
- max-norm regularization
- early stopping

3.7 pre-training
- stacks of RBMs
- autoencoders
- DBM
